Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIDeclarativeNetRequestCocoa.mm

```diff
@@ -47,6 +51,128 @@
 }
 
+void WebExtensionContext::loadDeclarativeNetRequestRulesetStateFromStorage()
+{
+    NSDictionary *savedRulesetState = objectForKey<NSDictionary>(m_state, declarativeNetRequestRulesetStateKey);
+    for (NSString *savedIdentifier in savedRulesetState) {
+        for (auto& ruleset : extension().declarativeNetRequestRulesets()) {
+            if (ruleset.rulesetID == String(savedIdentifier)) {
+                ruleset.enabled = objectForKey<NSNumber>(savedRulesetState, savedIdentifier).boolValue;
+                break;
+            }
+        }
+    }
+}
+
+void WebExtensionContext::saveDeclarativeNetRequestRulesetStateToStorage(NSDictionary *rulesetState)
+{
+    NSDictionary *savedRulesetState = objectForKey<NSDictionary>(m_state, declarativeNetRequestRulesetStateKey);
+    NSMutableDictionary *updatedRulesetState = savedRulesetState ? [savedRulesetState mutableCopy] : [NSMutableDictionary dictionary];
+
+    [updatedRulesetState addEntriesFromDictionary:rulesetState];
+
+    [m_state setObject:[updatedRulesetState copy] forKey:declarativeNetRequestRulesetStateKey];
+    writeStateToStorage();
+}
+
+void WebExtensionContext::clearDeclarativeNetRequestRulesetState()
+{
+    [m_state removeObjectForKey:declarativeNetRequestRulesetStateKey];
+}
+
+WebExtensionContext::DeclarativeNetRequestValidatedRulesets WebExtensionContext::declarativeNetRequestValidateRulesetIdentifiers(const Vector<String>& rulesetIdentifiers)
+{
+    WebExtension::DeclarativeNetRequestRulesetVector validatedRulesets;
+
+    WebExtension::DeclarativeNetRequestRulesetVector rulesets = extension().declarativeNetRequestRulesets();
+
+    for (auto identifier : rulesetIdentifiers) {
+        bool found = false;
+        for (auto ruleset : rulesets) {
+            if (ruleset.rulesetID == identifier) {
+                validatedRulesets.append(ruleset);
+                found = true;
+                break;
+            }
+        }
+
+        if (!found)
+            return { std::nullopt, toErrorString(@"declarativeNetRequest.updateEnabledRulesets", nil, @"Failed to apply rules. Invalid ruleset id: %@.", (NSString *)identifier) };
+    }
+
+    return { validatedRulesets, std::nullopt };
+}
+
+size_t WebExtensionContext::declarativeNetRequestEnabledRulesetCount()
+{
+    size_t count = 0;
+
+    for (auto& ruleset : extension().declarativeNetRequestRulesets()) {
+        if (ruleset.enabled)
+            ++count;
+    }
+
+    return count;
+}
+
+void WebExtensionContext::declarativeNetRequestToggleRulesets(const Vector<String>& rulesetIdentifiers, bool newValue, NSMutableDictionary *rulesetIdentifiersToEnabledState)
+{
+    for (auto& identifier : rulesetIdentifiers) {
+        for (auto& ruleset : extension().declarativeNetRequestRulesets()) {
+            if (ruleset.rulesetID != identifier)
+                continue;
+
+            ruleset.enabled = newValue;
+            rulesetIdentifiersToEnabledState[(NSString *)ruleset.rulesetID] = @(newValue);
```

Comment:
The `NSString` cast might not be needed. It is automatic in most cases.

Reply:
I think it wasn't automatic with the dictionary literal syntax when I first wrote it, but I changed it to use setObject:forKey: and I was able to get rid of the cast.